File: src/leetcode/python/CountNumbersWithUniqueDigits_357.py
class CountNumbersWithUniqueDigits_357(object):
    def countNumbersWithUniqueDigits(self, n):
        """
        f(i): if n = i, what is the unique digits number of 10^i
        f(0): 1
        f(1): f(0) + 9
        f(2): f(1) + 9 * 9
        f(3): f(2) + 9 * 9 * 8
        f(4): f(3) + 9 * 9 * 8 * 7
        ...
        ...
        ...
        f(n) = f(n - 1)(leading zero) + 9(non-leading zero) * remaining possibility

        final just sum the dp[0:n]
        """
        if not n:
            return 1
        dp = [0] * (n)
        dp[0] = 9
        for i in range(1, n):
            dp[i] = dp[i - 1] * (10 - i)
        return sum(dp) + 1

File: src/leetcode/java/CountNumbersWithUniqueDigits_357.java
package leetcode.java;

public class CountNumbersWithUniqueDigits_357 {
  // Stores:
  // [0]: 9,
  // [1]: 9 * 8,
  // [2]: 9 * 8 * 7,
  // [3]: 9 * 8 * 7 * 6,
  // [4]: 9 * 8 * 7 * 6 * 5,
  // [5]: 9 * 8 * 7 * 6 * 5 * 4,
  // [6]: 9 * 8 * 7 * 6 * 5 * 4 * 3,
  // [7]: 9 * 8 * 7 * 6 * 5 * 4 * 3 * 2,
  // [8]: 9 * 8 * 7 * 6 * 5 * 4 * 3 * 2 * 1,
  // [9]: 0,
  private static int[] uniques = new int[10];
  static {
    int num = 9;
    uniques[0] = 9;
    for (int i = 1; i < uniques.length; i++) {
      uniques[i] = uniques[i - 1] * (--num);
    }
    uniques[9] = 0;
  }

  public int countNumbersWithUniqueDigits(int n) {
    int[] dp = new int[n + 1];
    dp[0] = 1;
    for (int i = 1; i <= n; i++) {
      if (i > 10) break;
      // dp[i] = dp[i - 1](leading zero) + 9(non leading zero) * (remaining possibility)9 * 8 * .. 
      dp[i] = dp[i - 1] + uniques[0] * (i - 2 > 0 ? uniques[i - 2] : 1);
    }
    if (n > 10) return dp[10];
    return dp[n];
  }
}
